packages/notebook/src/panel.ts

```diff
@@ -78,40 +56,46 @@ const DIRTY_CLASS = 'jp-mod-dirty';
  * kernel on the context.
  */
 export
-class NotebookPanel extends Widget implements DocumentRegistry.IReadyWidget {
+class NotebookPanel extends DocumentWidget<Notebook, INotebookModel> {
   /**
    * Construct a new notebook panel.
    */
-  constructor(options: NotebookPanel.IOptions) {
-    super();
+  constructor(options: DocumentWidget.IOptions<Notebook, INotebookModel>) {
+    // Set default options
+    const notebookReady = new PromiseDelegate<void>();
+    options.ready = Promise.all([options.ready, notebookReady.promise]);
+    super(options);
+    this._activated = new Signal<this, void>(this);
+
+    // Set up CSS classes
     this.addClass(NOTEBOOK_PANEL_CLASS);
-    this.rendermime = options.rendermime;
-    let contentFactory = this.contentFactory = (
-      options.contentFactory || NotebookPanel.defaultContentFactory
-    );
-
-    let layout = this.layout = new BoxLayout();
-
-    // Toolbar
-    let toolbar = new Toolbar();
-    toolbar.addClass(NOTEBOOK_PANEL_TOOLBAR_CLASS);
-
-    // Notebook
-    let nbOptions: Notebook.IOptions = {
-      rendermime: this.rendermime,
-      languagePreference: options.languagePreference,
-      contentFactory: contentFactory,
-      mimeTypeService: options.mimeTypeService,
-      editorConfig: options.editorConfig,
-    };
-    let notebook = this.notebook = contentFactory.createNotebook(nbOptions);
-    notebook.addClass(NOTEBOOK_PANEL_NOTEBOOK_CLASS);
-
-    layout.addWidget(toolbar);
-    layout.addWidget(this.notebook);
-
-    BoxLayout.setStretch(toolbar, 0);
-    BoxLayout.setStretch(this.notebook, 1);
+    this.toolbar.addClass(NOTEBOOK_PANEL_TOOLBAR_CLASS);
+    this.content.addClass(NOTEBOOK_PANEL_NOTEBOOK_CLASS);
+
+    // Set up things related to the context
+    this.content.model = this.context.model;
+    // TODO: what happens when the context model changes? Can the context model change?
```

Comment:
I think we said the context model couldn't change

Reply:
Right, thanks.